### PR TITLE
KAA-1168 Add a page on Events

### PR DESCRIPTION
**Pull request by nocs00**

### CODE REVIEW

**Comment by kaa-leeroy** [comment]
Can one of the admins verify this patch?

**Comment by ashvayka** [comment]
This PR is incomplete, there is no "Kaa Events SDK API" section. This section should contain SDK APIs description similar to other sibling pages (profiling, grouping, configuration). Please take a look at corresponding PRs

File: doc/Programming-guide/Key-platform-features/Events/index.md [hunk 1]
```diff
@@ -4,4 +4,206 @@ title: Events
...
+* Enforcement of data integrity and validity
+* Efficient targeting of event recipients
+* Efficient and compact serialization
+
+It is the developer's responsibility to design event class schemas and make the client application interpret event data supplied by the endpoint library. The Kaa administrator, in turn, can provision those schemas into the Kaa server and generate the endpoint SDK.
+
+
+## Event generation and handling overview
+
+![](images/EventGeneration2.png)
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Events/index.md]: Please add some details. For example, "process of event generation and handling described on following diagram". Move "It is the developer's responsibility to design event class schemas and make the client application interpret event data supplied by the endpoint library. The Kaa administrator, in turn, can provision those schemas into the Kaa server and generate the endpoint SDK." to this section below the diagram. Add smth about scope of event exchange (user ids) and link to Owner verifiers page

**Comment by Oldnickus** [comment]
@kaa-leeroy 
add to whitelist

File: doc/Programming-guide/Key-platform-features/Events/index.md [hunk 2]
```diff
@@ -4,4 +4,917 @@ title: Events
...
+With the first sync request, the endpoint attempts to synchronize its event sequence number with the one stored at the Operation server. The server answers with either the sequence number of the latest event received from the endpoint or the number zero (if no events were received so far). If the number provided by the server differs from the number stored at the endpoint, the endpoint accepts the former and uses it as a starting number for new events.
+
+
+## SDK generation
+
+During the SDK generation, the Control server generates the event object model and extends the APIs to support methods for sending events and registering event listeners. The generated SDK can support multiple ECFs, although it cannot simultaneously support multiple versions of the same ECF.
+
+## Kaa Events SDK API
+
+#### Attach endpoint to user
```
Comment (sashadidukh) [inline doc/Programming-guide/Key-platform-features/Events/index.md:215]: Attach endpoint to owner description should be done in [this](http://jira.kaaproject.org/browse/KAA-1165) ticket.
Here we need to add a link to that description.

**Comment by ashvayka** [comment]
Please add Admin UI screen-shoots and REST API links to ECF provisioning, ECF mapping, etc.
Otherwise - LGTM

**Comment by nocs00** [comment]
Hi Andrew, please mention that REST API link provided in Kaa Events REST API .
Precise links couldn't be provided yet because the page with Server REST API description is empty in develop branch

**Comment by balldir** [comment]
@nocs00 , Please add todo for REST API links and create a ticket for updating this links when "Rest API" page will be ready. It will be possible to merge this pr after this.

**Comment by nocs00** [comment]
Task created http://jira.kaaproject.org/browse/KAA-1238 .
Todo links added .

**Comment by Slash32** [comment]
Looks good to me

File: doc/Programming-guide/Key-platform-features/Events/index.md [hunk 3]
```diff
@@ -4,4 +4,854 @@ title: Events
...
+
+## Event class and event class schema
+
+Each event is based on a particular event class (EC) that is defined by the corresponding event class schema. An EC is uniquely identified by a fully qualified name (FQN) and a tenant. In other words, there can be no two ECs with the same FQN within a single tenant.
+
+An event class schema format is based on the [Avro schema](http://avro.apache.org/docs/current/spec.html#schemas) with the additional attribute classType that supports two values: event and object. Kaa uses the classType attribute to distinguish actual events from objects, which are reusable parts of events. This is useful for avoiding redundant methods in SDK API.
+
+The following examples illustrate basic event class schemas.
+
+* The simplest definition of an event with the com.company.project.SimpleEvent FQN and no data fields
```
Comment (Polinaa) [inline doc/Programming-guide/Key-platform-features/Events/index.md:40]: SimpleEvent1

File: doc/Programming-guide/Key-platform-features/Events/index.md [hunk 4]
```diff
@@ -4,4 +4,854 @@ title: Events
...
+To send one or more events, the client should proceed as described in this section.
+
+**Get endpoint addresses**
+
+Execute the asynchronous findEventListeners method to request a list of the endpoints supporting all specified EC FQNs (FQN stands for fully qualified name).
+
+<ul class="nav nav-tabs">
+  <li class="active"><a data-toggle="tab" href="#java3">Java</a></li>
+  <li><a data-toggle="tab" href="#cpp3">C++</a></li>
+  <li><a data-toggle="tab" href="#c3">C</a></li>
```
Comment (Polinaa) [inline doc/Programming-guide/Key-platform-features/Events/index.md:312]: It'll be more comfortable to click on `C` tab if it is wider.

File: doc/Programming-guide/Key-platform-features/Events/index.md [hunk 5]
```diff
@@ -4,4 +4,854 @@ title: Events
...
+
+1. Open the **Event class families** window by clicking the corresponding link on the navigation panel.
+2. In the **Event class families** window, click **Add ECF**. 
+![](images/admin_ui/event_class_family/ecf1.png)
+3. In the **Add ECF** window, fill in all the required fields and then click **Add**.  
+**NOTE:** _the namespace and class name values should be unique._
+![](images/admin_ui/event_class_family/ecf2.png)
+4. In the **Event class family** window, add (optionally) an ECF schema by clicking **Add schema** under the **Schemas** table. 
+![](images/admin_ui/event_class_family/ecf3.png)
+5. In the **Add event class family schema** window, create an ECF schema either by using the **Event class family schema** schema form or by uploading the schema from a file, then click **Add**.
```
Comment (Polinaa) [inline doc/Programming-guide/Key-platform-features/Events/index.md:835]: Should **Add event class family schema** window be modified in Admin UI, as it has empty fields: _Version_, _Created by_ and _Date/Time created_?

File: doc/Programming-guide/Key-platform-features/Events/index.md (hunk 4 above)
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Events/index.md:312]: Please, report the issue to the designer. I believe, the design should ensure minimal tab width. Correct me if I'm wrong.

**Pull request closed without merging**